backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_multibit.cuh

```diff
@@ -81,57 +86,52 @@ __global__ void device_multi_bit_programmable_bootstrap_keybundle(
         bootstrapping_key, 0, rev_lwe_iteration, glwe_id, level_id,
         grouping_factor, 2 * polynomial_size, glwe_dimension, level_count);
-    const Torus *bsk_poly = bsk_slice + poly_id * params::degree;
+    const Torus *bsk_poly_ini = bsk_slice + poly_id * params::degree;
 
-    copy_polynomial<Torus, params::opt, params::degree / params::opt>(
-        bsk_poly, accumulator);
+    Torus reg_acc[params::opt];
 
-    // Accumulate the other terms
-    for (int g = 1; g < (1 << grouping_factor); g++) {
+    copy_polynomial_in_regs<Torus, params::opt, params::degree / params::opt>(
+        bsk_poly_ini, reg_acc);
 
-      const Torus *bsk_slice = get_multi_bit_ith_lwe_gth_group_kth_block(
-          bootstrapping_key, g, rev_lwe_iteration, glwe_id, level_id,
-          grouping_factor, 2 * polynomial_size, glwe_dimension, level_count);
-      const Torus *bsk_poly = bsk_slice + poly_id * params::degree;
+    int offset = get_start_ith_ggsw_offset(2 * polynomial_size, glwe_dimension,
+                                           level_count);
 
-      // Calculates the monomial degree
+    // Precalculate the monomial degrees and store them in shared memory
+    uint32_t *monomial_degrees = (uint32_t *)selected_memory;
+    if (threadIdx.x < (1 << grouping_factor)) {
       const Torus *lwe_array_group =
           block_lwe_array_in + rev_lwe_iteration * grouping_factor;
-      uint32_t monomial_degree = calculates_monomial_degree<Torus, params>(
-          lwe_array_group, g, grouping_factor);
-
-      synchronize_threads_in_block();
-      // Multiply by the bsk element
-      polynomial_accumulate_monic_monomial_mul<Torus>(
-          accumulator, bsk_poly, monomial_degree, threadIdx.x, params::degree,
-          params::opt, false);
+      monomial_degrees[threadIdx.x] = calculates_monomial_degree<Torus, params>(
+          lwe_array_group, threadIdx.x, grouping_factor);
     }
+    __syncthreads();
 
-    synchronize_threads_in_block();
+    // Accumulate the other terms
+    for (int g = 1; g < (1 << grouping_factor); g++) {
 
-    // Move accumulator to local memory
-    double2 temp[params::opt / 2];
-    int tid = threadIdx.x;
-#pragma unroll
-    for (int i = 0; i < params::opt / 2; i++) {
-      temp[i].x = __ll2double_rn((int64_t)accumulator[tid]);
-      temp[i].y =
-          __ll2double_rn((int64_t)accumulator[tid + params::degree / 2]);
-      temp[i].x /= (double)std::numeric_limits<Torus>::max();
-      temp[i].y /= (double)std::numeric_limits<Torus>::max();
-      tid += params::degree / params::opt;
+      uint32_t monomial_degree = monomial_degrees[g];
+
+      const Torus *bsk_poly = bsk_poly_ini + g * offset;
+      // Multiply by the bsk element
+      polynomial_product_accumulate_by_monomial_nosync<Torus, params>(
+          reg_acc, bsk_poly, monomial_degree);
     }
+    __syncthreads(); // need to make sure that the monomial_degrees are not
```

Comment:
> Here I don't understand why we need to synctreads again

Reply:
I was being extra careful, if you see in the following lines i reuse the shared memory to store the fft elements. I have seen that the compiler at low level adds a sync when I re-cast the shared memory, but i was afraid that this might not be the case for all architectures, so I thought it was better to sync explicitly there. I re-run the benchmarks after adding this sync (which was not there in the beginning) and the performance was not affected (probably because it was already done implicitly).

Reply:
Oh ok, maybe it would be worth updating the comment then to reflect this. It would mean that if the FFT takes data from the register directly we could avoid this synchronization?

Reply:
we would transfer this synchronization up to the calculation of the first fft level, but that requires to have a specific fft_direct that receives the data from registers, because in the other parts that call it the fft elements are expected in shared memory already (pbs or keygen).

Reply:
Yes totally, @bbarbakadze maybe something to look at with you next week at the office :slightly_smiling_face: